core/client/fs/src/test/java/alluxio/client/file/FileSystemFactoryTest.java

```diff
@@ -0,0 +1,137 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.security.Principal;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import javax.security.auth.Subject;
+
+public class FileSystemFactoryTest {
+
+  @Rule
+  public ExpectedException mThrown = ExpectedException.none();
+
+  @Before
+  public void before() {
+    ConfigurationUtils.reloadProperties();
+  }
+
+  @Test
+  public void testCloseRemovesFromCache() throws Exception {
+    FileSystem fs1 = FileSystem.Factory.get();
+    fs1.close();
+    assertTrue("FileSystem should be marked as closed", fs1.isClosed());
+    FileSystem fs2 = FileSystem.Factory.get();
+    assertFalse("FileSystem shouldn't be closed", fs2.isClosed());
+    assertNotSame("Should have different references", fs1, fs2);
+  }
+
+  @Test
+  public void singleMasterClientCacheTest()  {
+    clientCacheTest();
+  }
+
+  @Test
+  public void multiMasterClientCacheTest()  {
+    try (Closeable p = new SystemPropertyRule(PropertyKey.MASTER_RPC_ADDRESSES.getName(),
+        "192.168.0.1:1234,192.168.0.2:1445,192.168.0.3:9943").toResource()) {
+      ConfigurationUtils.reloadProperties();
+      InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+      MasterInquireClient.ConnectDetails connectDetails =
+          MasterInquireClient.Factory.getConnectDetails(conf);
+      // Make sure we have a MultiMaster authority
+      assertTrue(connectDetails.toAuthority() instanceof MultiMasterAuthority);
+      clientCacheTest();
+    } catch (IOException e) {
+      fail("Unable to set system properties");
+    }
+  }
+
+  @Test
+  public void zkClientCacheTest()  {
+    Map<String, String> sysProps = new HashMap<>();
+    sysProps.put(PropertyKey.ZOOKEEPER_ENABLED.getName(), Boolean.toString(true));
+    sysProps.put(PropertyKey.ZOOKEEPER_ADDRESS.getName(), "zk@192.168.0.5");
+    sysProps.put(PropertyKey.ZOOKEEPER_ELECTION_PATH.getName(), "/leader");
+
+    try (Closeable p = new SystemPropertyRule(sysProps).toResource()) {
+      ConfigurationUtils.reloadProperties();
+      InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+      MasterInquireClient.ConnectDetails connectDetails =
+          MasterInquireClient.Factory.getConnectDetails(conf);
+      // Make sure we have a Zookeeper authority
+      assertTrue(connectDetails.toAuthority() instanceof ZookeeperAuthority);
+      clientCacheTest();
+    } catch (IOException e) {
+      fail("Unable to set system properties");
+    }
+  }
+
+  @Test
+  public void nullSubjectTest()  {
+    mThrown.expect(NullPointerException.class);
+    FileSystem.Factory.get(null);
+  }
+
+  public void clientCacheTest()  {
+    resetClientCache();
+    FileSystem fs1 = FileSystem.Factory.get();
+    FileSystem fs2 = FileSystem.Factory.get();
+    assertEquals("Second client should have been retrieved from cache.", fs1, fs2);
```

Comment:
Should we use `assertSame`/`assertNotSame` in this test? We might implement `equals` some day to compare filesystems by something other than reference

Reply:
Good catch. That's actually what I meant to use.